### PR TITLE
secp256k1: Optimize NAF conversion.

### PR DESCRIPTION
**Pull request by davecgh**
**This is rebased on #2690**.

This significantly optimizes the NAF conversion code by rewriting it to avoid all heap allocations as well as switch to an O(1) algorithm.

It also reworks the NAF tests to actually test the produced encoding adheres to the requires of NAF encoding as well as to make them more consistent with modern practices in the code.

Specifically, the following properties are asserted:

- There must not be a leading zero byte and the number of bytes used to encode the negative portion must not exceed the positive portion
- There must not be any adjacent non-zero digits
- The negative and positive portions must sum back to the original value

The existing tests only ensure the positive and negative portions sum back up to the original value.

The updated tests also revealed that the existing naf function doesn't strip leading zeroes, so that logic has been added.  That behavior is not a problem in terms of correctness given how it is used in the code since extra leading zeroes don't change the resulting value, but it could potentially cause more work to be done than necessary and having leading zeroes technically doesn't conform to NAF encoding.

Finally, it slightly reworks the scalar multiplication logic to improve its readability by making use of zero values and a single mask instead of shifting every individual byte of the positive and negative portions of k1 and k2.

The following benchmark shows a before and after comparison of the NAF conversion as well as how that translates to scalar multiplication and signature verification:

```
name        old time/op      new time/op     delta
----------------------------------------------------------------------
NAF          1.16µs ± 1%     0.08µs ± 1%     -93.02%  (p=0.008 n=5+5)
ScalarMult    138µs ± 1%      135µs ± 0%      -1.77%  (p=0.016 n=5+4)
SigVerify     164µs ± 0%      162µs ± 0%      -0.98%  (p=0.008 n=5+5)

name         old alloc/op    new alloc/op    delta
----------------------------------------------------------------------
NAF           96.0B ± 0%       0.0B          -100.00%  (p=0.008 n=5+5)
ScalarMult     816B ± 0%       720B ± 0%      -11.76%  (p=0.008 n=5+5)
SigVerify    1.54kB ± 0%     1.44kB ± 0%      -6.25%   (p=0.008 n=5+5)

name         old allocs/op   new allocs/op   delta
----------------------------------------------------------------------
NAF          2.00 ± 0%       0.00            -100.00%  (p=0.008 n=5+5)
ScalarMult   15.0 ± 0%       11.0 ± 0%        -26.67%  (p=0.008 n=5+5)
SigVerify    32.0 ± 0%       28.0 ± 0%        -12.50%  (p=0.008 n=5+5)
```

### CODE REVIEW

File: dcrec/secp256k1/curve.go [hunk 1]
```diff
@@ -636,6 +639,142 @@ func splitK(k []byte) ([]byte, []byte, int, int) {
...
+	// The following code makes use of that formula to compute the NAF more
+	// efficiently.
+	//
+	// To understand the logic here, observe that the only way the NAF has a
+	// nonzero digit at a given bit is when either 3k/2 or k/2 has a bit set in
+	// that position, but not both.  In other words, the result of a bitwise
+	// xor.  This can be seen simply by considering that when the bits are the
+	// same, the subtraction is either 0-0 or 1-1, both of which are 0.
+	//
+	// Further, observe that the "1" digits in the result are contributed by
```
Comment (davecgh) [inline dcrec/secp256k1/curve.go:735]:
```suggestion
	// Further, observe that the "+1" digits in the result are contributed by
```

Comment (davecgh) [reply dcrec/secp256k1/curve.go]: Updated.

**Comment by matheusd** [comment]
Commit msg for [Improve scalar mult readability](https://github.com/decred/dcrd/pull/2695/commits/7216d26a59a328dc9ccd3a2c068de7ea446f4ab5) has "might not otherwise **by** obvious"

File: dcrec/secp256k1/curve_test.go [hunk 2]
```diff
@@ -527,6 +527,120 @@ func TestDoubleAffine(t *testing.T) {
...
+	}
+	if len(neg) > len(pos) {
+		return fmt.Errorf("negative has len %d > pos len %d", len(neg),
+			len(pos))
+	}
+
+	// Ensure the result doesn't have any adjacent non-zero digits.
+	gotPos := new(big.Int).SetBytes(pos)
+	gotNeg := new(big.Int).SetBytes(neg)
+	posOrNeg := new(big.Int).Or(gotPos, gotNeg)
```
[the remark this answers is not in the document]
Comment (davecgh) [reply dcrec/secp256k1/curve_test.go:547]: Right, NAF is specifically the unique representation where there are no adjacent (side-by-side) non-zero digits, where those digits are from the set {0,±1}.  In other words, there will always be a zero on either side of every digit (where those digits are +1 or -1).

I suspect it's probably easier to change the notation slightly to say that `p` represents the `+1` signed digit and `n` represents the `-1` signed digit, and `0` represents the `0` signed digit.  Every number will always be some combination of `...{n,p}0{n,p}0{n,p}0...`.

Taking the `n = 478` case as an example, the NAF in this notation is `p000n000n0`.  That is `2^9 - 2^5 - 2 = 512 - 32 - 2 = 478`.  

When split up into the positive and negative arrays, it then would be:

```
pos: 1000000000  <--- all of the `p` (aka `+1`) digits have a 1 bit set here
neg: 0000100010  <--- all of the `n` (aka `-1`) digits have a 1 bit set here
```

Thus, taking their bitwise or will end up with bits set in the position where there is either a `p` (aka `+1`) or an `n` (aka `-1`) which means that any two adjacent bits being set implies there were two adjacent non-zero signed digits, which is a violation of the encoding.

**Comment by davecgh** [comment]
> 
> 
> Commit msg for [Improve scalar mult readability](https://github.com/decred/dcrd/pull/2695/commits/7216d26a59a328dc9ccd3a2c068de7ea446f4ab5) has "might not otherwise **by** obvious"

Corrected.

**Review by matheusd — APPROVED** [review]

**Comment by davecgh** [comment]
I added an additional edge-condition test case for which results in a 257-bit NAF encoding since I noticed that wasn't explicitly tested.

**Review by rstaudt2 — APPROVED** [review]

**Review by JoeGruffins — APPROVED** [review]
